utcp.c

```diff
@@ -377,6 +386,61 @@ uint32_t buffer_free(const struct buffer *buf) {
     return buf->maxsize - buf->used;
 }
 
+static void utcp_send_error(const struct pkt_t *pkt, size_t len, ssize_t sent, bool drop) {
+    if(sent != len) {
+        if(sent > len) {
+            debug("Error: sent packet %u and ack %u but with a larger size than it should, %u of %u bytes sent", pkt->hdr.seq, pkt->hdr.ack, sent, len);
+        }
+        else if(sent >= 0) {
+            // we do not handle split packets
+            debug("Warning: failed to send packet %u and ack %u with only %u of %u bytes packet size sent, %s", pkt->hdr.seq, pkt->hdr.ack, sent, len, drop? "dropping the packet" : "retrying it later");
+        }
+        else if(sent == UTCP_WOULDBLOCK) {
+            debug("Debug: failed to send packet %u and ack %u with UTCP_WOULDBLOCK, %s", pkt->hdr.seq, pkt->hdr.ack, drop? "dropping the packet" : "retrying it later");
+        }
+        else {
+            // the pkt receiver might have gone offline causing the routing to fail
+            // drop the packet and continue
+            debug("Error: failed to send packet %u and ack %u with error %u, %s", pkt->hdr.seq, pkt->hdr.ack, sent, drop? "dropping the packet" : "retrying it later");
+        }
+    }
+    else if(drop) {
+        debug("Error: failed to send packet %u and ack %u, dropping the packet [sent=%u]", pkt->hdr.seq, pkt->hdr.ack, sent);
+    }
+}
+
+static bool utcp_send_packet_or_queue(struct utcp *utcp, const struct pkt_t *pkt, size_t len) {
+    ssize_t sent = utcp->send(utcp, pkt, len);
```

Comment:
I overworked this so packets now are queued when there already are packets pending and to not further queue packets on retransmit
the pending packets are always dealt with first, to be followed by the retransmit and finally the polling and ack
thereby I also fixed to send already buffered data when there's nothing to poll
I think this should be fine. on the one hand polling, ack and retransmit always attempts to (re)send *new* data from the buffer and so should be in order with the pending packets sent from the queue when sent last. on the other hand there only are unbuffered packets sent left for the SYN and RST on connection establishment and reset and utcp should be robust from packet reordering anyway.